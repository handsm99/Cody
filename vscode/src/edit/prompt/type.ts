export interface LLMInteraction {
    prompt: string
    stopSequences?: string[]
    responseTopic?: string
    assistantText?: string
    assistantPrefix?: string
}

export interface GetLLMInteractionOptions {
    instruction: string
    precedingText: string
    selectedText: string
    followingText: string
    fileName: string
}

type LLMInteractionBuilder = (options: GetLLMInteractionOptions) => LLMInteraction

export interface EditLLMInteraction {
    getEdit: LLMInteractionBuilder
    getDoc: LLMInteractionBuilder
    getFix: LLMInteractionBuilder
    getAdd: LLMInteractionBuilder
}
