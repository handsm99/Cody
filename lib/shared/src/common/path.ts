import type { URI } from 'vscode-uri'

import { isWindows as _isWindows } from './platform'

export interface PathFunctions {
    /**
     * All but the last element of path, or "." if that would be the empty path.
     */
    dirname: (path: string) => string

    /**
     * The last element of path, or "" if path is empty.
     * @param path the path to operate on
     * @param suffix optional suffix to remove
     */
    basename: (path: string, suffix?: string) => string

    /** The extension of path, including the last '.'. */
    extname: (path: string) => string

    /** Path separator. */
    separator: string
}

/** For file system paths on Windows ('\' separators and drive letters). */
export const windowsFilePaths: PathFunctions = pathFunctions(true)

/**
 * For POSIX and URI paths ('/' separators).
 */
export const posixAndURIPaths: PathFunctions & {
    /**
     * The relative path from {@link from} to {@link to}.
     *
     * Only implemented for POSIX and URI paths because there are currently no callers that need
     * this for Windows paths.
     */
    relative: (from: string, to: string) => string
} = { ...pathFunctions(false), relative: posixAndURIPathsRelative }

/**
 * Get the {@link PathFunctions} to use for the given URI's path.
 */
export function pathFunctionsForURI(uri: URI, isWindows = _isWindows()): PathFunctions {
    return uri.scheme === 'file' && isWindows ? windowsFilePaths : posixAndURIPaths
}

// I don't like reimplementing this here, but it's the best option because: (1) using Node's `path`
// module requires us to configure the bundler right or risk a subtle bug (and path-browserify
// doesn't have `path/win32` support, so we'd need multiple underlying packages); and (2)
// `vscode-uri` is hard to test because it has a global constant for the current platform set at
// init time.
function pathFunctions(isWindows: boolean): PathFunctions {
    const sep = isWindows ? '\\' : '/'
    const f: PathFunctions = {
        dirname(path: string): string {
            if (path === '') {
                return '.'
            }
            if (isWindows && isDriveLetter(path)) {
                return path
            }
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            if (path === '') {
                return sep
            }
            const i = path.lastIndexOf(sep)
            if (i === -1) {
                return '.'
            }
            if (i === 0) {
                return sep
            }
            path = path.slice(0, i)
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            return path
        },
        basename(path: string, suffix?: string): string {
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return ''
            }
            path = path.split(sep).at(-1) ?? ''
            if (suffix && path.endsWith(suffix)) {
                path = path.slice(0, -suffix.length)
            }
            return path
        },
        extname(path: string): string {
            const basename = f.basename(path)
            const i = basename.lastIndexOf('.')
            if (i === 0 || i === -1) {
                return ''
            }
            return basename.slice(i)
        },
        separator: sep,
    }
    return f
}

function isDriveLetter(path: string): boolean {
    return /^[A-Za-z]:$/.test(path)
}

/** The relative path from {@link from} to {@link to}. */
function posixAndURIPathsRelative(from: string, to: string): string {
    // Normalize slashes.
    from = from.replaceAll(/\/{2,}/g, '/')
    to = to.replaceAll(/\/{2,}/g, '/')

    // Trim trailing slashes.
    if (from !== '/' && from.endsWith('/')) {
        from = from.slice(0, -1)
    }
    if (to !== '/' && to.endsWith('/')) {
        to = to.slice(0, -1)
    }

    if (from === to) {
        return ''
    }

    // From a relative to an absolute path, the absolute path dominates.
    if (!from.startsWith('/') && to.startsWith('/')) {
        return to
    }

    const fromParts = from === '/' ? [''] : from.split('/')
    const toParts = to === '/' ? [''] : to.split('/')

    // Find the common root.
    let commonLength = 0
    while (
        commonLength < fromParts.length &&
        commonLength < toParts.length &&
        fromParts[commonLength] === toParts[commonLength]
    ) {
        commonLength++
    }

    // Calculate the number of directory traversals needed.
    const traversals = fromParts.length - commonLength

    // Build the relative path.
    const relativePath: string[] = []
    for (let i = 0; i < traversals; i++) {
        relativePath.push('..')
    }

    // Add the non-common path parts from the 'to' path.
    relativePath.push(...toParts.slice(commonLength))

    return relativePath.join('/')
}
