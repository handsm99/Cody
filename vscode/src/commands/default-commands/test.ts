import type { ContextFile } from '@sourcegraph/cody-shared'
import { getEditor } from '../../editor/active-editor'
import { getContextFileFromCursor } from '../context/selection'
import { getContextFilesForTests } from '../context/test-files'
import type { ExecuteChatArguments } from '.'
import type { ChatSession } from '../../chat/chat-view/SimpleChatPanelProvider'
import { executeChat } from './ask'

/**
 * Generates the prompt and context files with arguments for the 'test' command.
 *
 * Context: Test files, current selection, and current file
 */
export async function testCommand(): Promise<{ prompt: string; args: ExecuteChatArguments }> {
    const prompt =
        "Review the shared code context and configurations to identify the test framework and libraries in use. Then, generate a suite of multiple unit tests for the functions in <selected> using the detected test framework and libraries. Be sure to import the function being tested. Follow the same patterns as any shared context. Only add packages, imports, dependencies, and assertions if they are used in the shared code. Pay attention to the file path of each shared context to see if test for <selected> already exists. If one exists, focus on generating new unit tests for uncovered cases. If none are detected, import common unit test libraries for this programming language. Focus on validating key functionality with simple and complete assertions. Only include mocks if one is detected in the shared code. Before writing the tests, identify which test libraries and frameworks to import, e.g. 'No new imports needed - using existing libs' or 'Importing test framework that matches shared context usage' or 'Importing the defined framework', etc. Then briefly summarize test coverage and any limitations. At the end, enclose the full completed code for the new unit tests, including all necessary imports, in a single markdown codeblock. No fragments or TODO. The new tests should validate expected functionality and cover edge cases for <selected> with all required imports, including importing the function being tested. Do not repeat existing tests."

    const editor = getEditor()?.active
    const document = editor?.document
    const contextFiles: ContextFile[] = []

    if (document) {
        try {
            const cursorContext = await getContextFileFromCursor()
            if (cursorContext) {
                contextFiles.push(cursorContext)
            }

            contextFiles.push(...(await getContextFilesForTests(document.uri)))
        } catch (error) {
            console.error(error)
        }
    }

    return {
        prompt,
        args: {
            userContextFiles: contextFiles,
            addEnhancedContext: false,
            source: 'test',
        },
    }
}

/**
 * Executes the text command as a chat command via 'cody.action.chat'
 */
export async function executeTestCommand(): Promise<ChatSession | undefined> {
    const { prompt, args } = await testCommand()
    return executeChat(prompt, args)
}
