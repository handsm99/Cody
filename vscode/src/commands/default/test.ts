import { logError, type ContextFile, logDebug } from '@sourcegraph/cody-shared'
import { getEditor } from '../../editor/active-editor'
import { getContextFileFromCursor } from '../context/selection'
import type { CodyCommandArgs } from '../types'
import { type ExecuteChatArguments, executeChat } from './ask'
import { defaultCommands } from '.'
import type { ChatCommandResult } from '../../main'
import { getContextFilesForTestCommand } from '../context/test-command'
/**
 * Generates the prompt and context files with arguments for the 'test' command.
 *
 * Context: Test files, current selection, and current file
 */
async function testCommand(): Promise<ExecuteChatArguments> {
    const prompt = defaultCommands.test.prompt

    const editor = getEditor()?.active
    const document = editor?.document
    const contextFiles: ContextFile[] = []

    if (document) {
        try {
            const cursorContext = await getContextFileFromCursor()
            contextFiles.push(...cursorContext)

            contextFiles.push(...(await getContextFilesForTestCommand(document.uri)))
        } catch (error) {
            logError('testCommand', 'failed to fetch context', { verbose: error })
        }
    }

    return {
        text: prompt,
        contextFiles,
        addEnhancedContext: false,
        source: 'test',
        submitType: 'user-newchat',
    }
}

/**
 * Executes the text command as a chat command via 'cody.action.chat'
 */
export async function executeTestCommand(
    args?: Partial<CodyCommandArgs>
): Promise<ChatCommandResult | undefined> {
    logDebug('executeTestCommand', 'executing', { args })
    return {
        type: 'chat',
        session: await executeChat(await testCommand()),
    }
}
