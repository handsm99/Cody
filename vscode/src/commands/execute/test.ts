import { logError, type ContextFile, logDebug } from '@sourcegraph/cody-shared'
import { getEditor } from '../../editor/active-editor'
import { getContextFileFromCursor } from '../context/selection'
import type { CodyCommandArgs } from '../types'
import { type ExecuteChatArguments, executeChat } from './ask'
import { defaultCommands } from '.'
import type { ChatCommandResult } from '../../main'
import { getContextFilesForTestCommand } from '../context/test-command'
import { telemetryService } from '../../services/telemetry'
import { telemetryRecorder } from '../../services/telemetry-v2'
/**
 * Generates the prompt and context files with arguments for the 'test' command.
 *
 * Context: Test files, current selection, and current file
 */
async function testCommand(args?: Partial<CodyCommandArgs>): Promise<ExecuteChatArguments> {
    let prompt = defaultCommands.test.prompt

    if (args?.additionalInstruction) {
        prompt = `${prompt} ${args.additionalInstruction}`
    }

    const editor = getEditor()?.active
    const document = editor?.document
    const contextFiles: ContextFile[] = []

    if (document) {
        try {
            const cursorContext = await getContextFileFromCursor()
            contextFiles.push(...cursorContext)

            contextFiles.push(...(await getContextFilesForTestCommand(document.uri)))
        } catch (error) {
            logError('testCommand', 'failed to fetch context', { verbose: error })
        }
    }

    return {
        text: prompt,
        contextFiles,
        addEnhancedContext: false,
        source: 'test',
        submitType: 'user-newchat',
    }
}

/**
 * Executes the text command as a chat command via 'cody.action.chat'
 */
export async function executeTestCommand(
    args?: Partial<CodyCommandArgs>
): Promise<ChatCommandResult | undefined> {
    logDebug('executeTestCommand', 'executing', { args })
    telemetryService.log('CodyVSCodeExtension:command:test:executed', {
        useCodebaseContex: false,
        requestID: args?.requestID,
        source: args?.source,
    })
    telemetryRecorder.recordEvent('cody.command.test', 'executed', {
        metadata: {
            useCodebaseContex: 0,
        },
        interactionID: args?.requestID,
        privateMetadata: {
            requestID: args?.requestID,
            source: args?.source,
        },
    })

    return {
        type: 'chat',
        session: await executeChat(await testCommand(args)),
    }
}
