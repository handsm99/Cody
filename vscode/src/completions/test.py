# Author: <NAME>
# Mail: <EMAIL>

# try to make it autocomplete add
def ad

def main():
    print("Hello World!")
    return 0
