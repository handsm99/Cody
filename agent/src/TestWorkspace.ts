import { execSync } from 'node:child_process'
import { mkdtempSync } from 'node:fs'
import fspromises from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'
import { DOTCOM_URL } from '@sourcegraph/cody-shared'
import { expect } from 'vitest'
import * as vscode from 'vscode'

/**
 * A test workspace is a temporary directory that is created before running tests.
 * Use the `beforeAll` and `afterAll` methods to set up and tear down the workspace.
 **/
export class TestWorkspace {
    public readonly rootUri: vscode.Uri
    public readonly rootPath: string

    constructor(public readonly prototypePath: string) {
        const tmpdir = mkdtempSync(path.join(os.tmpdir(), 'cody-vscode-shim-test'))
        this.rootUri = vscode.Uri.file(tmpdir)
        this.rootPath = tmpdir
    }

    public async beforeAll(): Promise<void> {
        const mayRecord =
            process.env.CODY_RECORDING_MODE === 'record' || process.env.CODY_RECORD_IF_MISSING === 'true'
        if (mayRecord) {
            try {
                execSync('src login', { stdio: 'inherit' })
            } catch {
                throw new Error(
                    "Can't record HTTP requests without being authenticated. " +
                        'To fix this problem, run:\n  source agent/scripts/export-cody-http-recording-tokens.sh'
                )
            }
            expect(new URL(process.env.SRC_ENDPOINT ?? '')).toStrictEqual(DOTCOM_URL)
        }

        await fspromises.mkdir(this.rootPath, { recursive: true })
        await fspromises.cp(this.prototypePath, this.rootPath, {
            recursive: true,
        })
    }
    public file(...pathParts: string[]): vscode.Uri {
        return vscode.Uri.file(path.join(this.rootPath, ...pathParts))
    }

    public async afterAll(): Promise<void> {
        await fspromises.rm(this.rootPath, {
            recursive: true,
            force: true,
        })
    }
}
