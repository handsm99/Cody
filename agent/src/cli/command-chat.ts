import ora, { spinners } from 'ora'

import path from 'node:path'

import type { Polly } from '@pollyjs/core'
import { type ContextItem, ModelUsage, TokenCounter } from '@sourcegraph/cody-shared'
import { Command } from 'commander'

import * as vscode from 'vscode'
import { activate } from '../../../vscode/src/extension.node'
import { startPollyRecording } from '../../../vscode/src/testutils/polly'
import packageJson from '../../package.json'
import { newEmbeddedAgentClient } from '../agent'
import type { ClientInfo } from '../protocol-alias'
import { Streams } from './Streams'
import { AuthenticatedAccount } from './command-auth/AuthenticatedAccount'
import { notLoggedIn } from './command-auth/messages'

declare const process: { pkg: { entrypoint: string } } & NodeJS.Process
export interface ChatOptions {
    endpoint: string
    accessToken: string
    message: string
    dir: string
    model?: string
    contextRepo?: string[]
    contextFile?: string[]
    showContext: boolean
    debug: boolean
    silent: boolean
    isTesting?: boolean
    streams?: Streams
}

export const chatCommand = () =>
    new Command('chat')
        .description('Chat with codebase context')
        .requiredOption('-m, --message <message>', 'Message to send')
        .option(
            '--endpoint',
            'Sourcegraph instance URL',
            process.env.SRC_ENDPOINT ?? 'https://sourcegraph.com'
        )
        .option(
            '--access-token',
            'Sourcegraph access token. ' + loginInstruction,
            process.env.SRC_ACCESS_TOKEN ?? ''
        )
        .option('-C, --dir <dir>', 'Run in directory <dir>', process.cwd())
        .option('--model <model>', 'Chat model to use')
        .option(
            '--context-repo <repos...>',
            '(Sourcegraph Enterprise only) Names of repositories to use as context. Example: github.com/sourcegraph/cody.'
        )
        .option('--context-file <files...>', 'Local files to include in the context')
        .option('--show-context', 'Show context items in reply', false)
        .option('--silent', 'Disable streaming reply', false)
        .option('--debug', 'Enable debug logging', false)
        .action(async (options: ChatOptions) => {
            if (!options.accessToken) {
                const spinner = ora().start('Loading access token')
                const account = await AuthenticatedAccount.fromUserSettings(spinner)
                if (!spinner.isSpinning) {
                    process.exit(1)
                }
                spinner.stop()
                if (account) {
                    options.accessToken = account.accessToken
                    options.endpoint = account.serverEndpoint
                }
            }
            let polly: Polly | undefined
            if (process.env.CODY_RECORDING_DIRECTORY && process.env.CODY_RECORDING_NAME) {
                polly = startPollyRecording({
                    recordingName: process.env.CODY_RECORDING_NAME,
                    recordingDirectory: process.env.CODY_RECORDING_DIRECTORY,
                })
            }
            let exitCode = 0
            try {
                exitCode = await chatAction(options)
            } finally {
                await polly?.stop()
            }
            process.exit(exitCode)
        })

const loginInstruction = 'Sign in with the command: cody auth login --web'

export async function chatAction(options: ChatOptions): Promise<number> {
    const streams = options.streams ?? Streams.default()
    const spinner = ora({
        spinner: spinners.fistBump,
        isSilent: options.silent,
        stream: streams.stderr,
    }).start()
    const workspaceRootUri = vscode.Uri.file(path.resolve(options.dir))
    const clientInfo: ClientInfo = {
        name: 'cody-cli',
        version: options.isTesting ? '0.1.0-SNAPSHOT' : packageJson.version,
        workspaceRootUri: workspaceRootUri.toString(),
        extensionConfiguration: {
            serverEndpoint: options.endpoint,
            accessToken: options.accessToken,
            customHeaders: {},
            customConfiguration: {
                'cody.experimental.symf.enabled': false,
                'cody.experimental.telemetry.enabled': options.isTesting ? false : undefined,
            },
        },
    }
    spinner.text = 'Initializing...'
    const { serverInfo, client, messageHandler } = await newEmbeddedAgentClient(clientInfo, activate)
    const { models } = await client.request('chat/models', { modelUsage: ModelUsage.Chat })

    messageHandler.registerNotification('webview/postMessage', message => {
        if (message.message.type === 'transcript') {
            const lastMessage = message.message.messages.at(-1)
            if (lastMessage?.model && !spinner.text.startsWith('Model')) {
                const modelName =
                    models.find(model => model.model === lastMessage.model)?.title ?? lastMessage.model
                spinner.text = modelName
                spinner.spinner = spinners.dots
            }
            spinner.prefixText = (lastMessage?.text ?? '') + '\n'
        }
    })

    if (!serverInfo.authStatus?.isLoggedIn) {
        notLoggedIn(spinner)
        return 1
    }

    spinner.text = 'Asking Cody...'
    const id = await client.request('chat/new', null)

    if (options.model) {
        void client.request('webview/receiveMessage', {
            id,
            message: {
                command: 'chatModel',
                model: options.model,
            },
        })
    }

    if (options.contextRepo && options.contextRepo.length > 0) {
        if (serverInfo.authStatus?.isDotCom) {
            spinner.fail(
                'The --context-repo option is only available for Sourcegraph Enterprise users. ' +
                    'Please sign into an Enterprise instance with the command: cody auth logout && cody auth login --web'
            )
            return 1
        }
        const { repos } = await client.request('graphql/getRepoIds', {
            names: options.contextRepo,
            first: options.contextRepo.length,
        })
        await client.request('webview/receiveMessage', {
            id,
            message: {
                command: 'context/choose-remote-search-repo',
                explicitRepos: repos,
            },
        })
    }

    const contextFiles: ContextItem[] = []
    for (const relativeOrAbsolutePath of options.contextFile ?? []) {
        contextFiles.push({
            type: 'file',
            uri: toUri(options.dir, relativeOrAbsolutePath),
        })
    }
    const start = performance.now()
    const response = await client.request('chat/submitMessage', {
        id,
        message: {
            command: 'submit',
            submitType: 'user',
            text: options.message,
            contextFiles,
            addEnhancedContext: false,
        },
    })

    if (response.type !== 'transcript') {
        spinner.fail(
            `Unexpected chat reply. Expected type "transcript", got "${response.type}"` +
                JSON.stringify(response, null, 2)
        )
        return 1
    }

    const reply = response.messages.at(-1)
    if (!reply) {
        spinner.fail(
            'Unexpected chat reply. Expected non-empty messages, got' + JSON.stringify(response, null, 2)
        )
        return 1
    }

    if (reply.error) {
        spinner.fail(`Unexpected error: ${JSON.stringify(reply, null, 2)}`)
        return 1
    }

    spinner.spinner = spinners.triangle
    spinner.prefixText = ''
    const elapsed = performance.now() - start
    const replyText = reply.text ?? ''
    const tokens = TokenCounter.encode(replyText).length
    const tokensPerSecond = tokens / (elapsed / 1000)
    spinner.text = spinner.text.trim() + ` (${Math.round(tokensPerSecond)} tokens/second)`
    spinner.clear()

    if (options.showContext) {
        const contextFiles = reply.contextFiles ?? []
        streams.log('> Context items:\n')
        for (const [i, item] of contextFiles.entries()) {
            streams.log(`> ${i + 1}. ${item.uri.fsPath}\n`)
        }
        streams.log('\n')
    }
    streams.log(replyText + '\n')
    await client.request('shutdown', null)
    spinner.succeed()
    return 0
}

function toUri(dir: string, relativeOrAbsolutePath: string): vscode.Uri {
    const absolutePath = path.isAbsolute(relativeOrAbsolutePath)
        ? relativeOrAbsolutePath
        : path.resolve(dir, relativeOrAbsolutePath)
    return vscode.Uri.file(absolutePath)
}
