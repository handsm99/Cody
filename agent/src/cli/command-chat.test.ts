import path from 'node:path'
import { describe } from 'node:test'
import type { Polly } from '@pollyjs/core'
import { afterAll, beforeAll, expect, it } from 'vitest'
import YAML from 'yaml'
import { startPollyRecording } from '../../../vscode/src/testutils/polly'
import { TESTING_CREDENTIALS } from '../../../vscode/src/testutils/testing-credentials'
import { buildAgentBinary, getAgentDir } from '../TestClient'
import { TestWorkspace } from '../TestWorkspace'
import { Streams, StringBufferStream } from './Streams'
import { isWindows } from './command-bench/isWindows'
import { type ChatOptions, chatAction, chatCommand } from './command-chat'

process.env.CODY_SHIM_TESTING = 'true'
process.env.DISABLE_FEATURE_FLAGS = 'true'

interface ChatCommandResult {
    command: string
    exitCode: number
    stdout: string
    stderr: string
}

describe('cody chat', () => {
    const credentials = TESTING_CREDENTIALS.s2
    let polly: Polly
    const agentDirectory = getAgentDir()
    const tmp = new TestWorkspace(path.join(agentDirectory, 'src', '__tests__', 'cody-cli-chat'))
    async function runCommand(params: {
        args: string[]
        expectedExitCode?: number
    }): Promise<ChatCommandResult> {
        // For some reason, I can't get the option parsing working with
        // `--access-token` or `--endpoint` so we modify process.env instead.
        process.env.SRC_ACCESS_TOKEN = credentials.token ?? credentials.redactedToken
        process.env.SRC_ENDPOINT = credentials.serverEndpoint
        process.env.DISABLE_FEATURE_FLAGS = 'true'
        process.env.CODY_TELEMETRY_EXPORTER = 'testing'
        const args = [...params.args, '--dir', tmp.rootPath, '--silent']

        const command = chatCommand()
        const parseResult = command.parseOptions(args)
        if (parseResult.unknown.length > 0) {
            throw new Error(`Unknown options: ${parseResult.unknown.join(', ')}`)
        }
        const options = command.opts<ChatOptions>()
        options.isTesting = true

        const stdout = new StringBufferStream()
        const stderr = new StringBufferStream()
        options.streams = new Streams(stdout, stderr)
        // Uncomment below to see output channel logs from Cody
        // options.debug = true
        const exitCode = await chatAction(options)
        if (exitCode !== (params.expectedExitCode ?? 0)) {
            const extraHint =
                stdout.buffer.length === 0 && stderr.buffer.length === 0
                    ? 'Stdout and stderr are empty even if the process exited with a non-zero code. ' +
                      'Comment out the --silent option from the test file to try to get more debugging information.'
                    : undefined
            throw new Error(
                YAML.stringify({
                    exitCode,
                    expectedExitCode: params.expectedExitCode,
                    stdout: stdout.buffer,
                    stderr: stderr.buffer,
                    extraHint,
                })
            )
        }
        return {
            command: 'cody chat ' + params.args.join(' '),
            exitCode,
            stdout: stdout.buffer.replaceAll(tmp.rootPath, 'WORKING_DIRECTORY'),
            stderr: stderr.buffer,
        }
    }

    beforeAll(() => {
        tmp.beforeAll()
        buildAgentBinary()
        polly = startPollyRecording({
            recordingName: 'cody-chat',
            keepUnusedRecordings: process.env.CODY_KEEP_UNUSED_RECORDINGS === 'true',
            recordingDirectory: path.join(agentDirectory, 'recordings'),
        })
    })

    afterAll(async () => {
        tmp.afterAll()
        await polly.stop()
    })

    it('--message (hello world test)', async () => {
        expect(
            YAML.stringify(
                await runCommand({
                    args: ['chat', '-m', 'respond with "hello" and nothing else'],
                })
            )
        ).toMatchSnapshot()
    }, 10_000)

    // This test is failing on macOS. It reports remote search results as failing
    // the context filter, probably because it does not wait for the context filter fetch.
    // This test is failing consistently on windows. https://linear.app/sourcegraph/issue/CODY-2912/cli-squirrel-test-failing-on-windows
    it.skip('--context-repo (squirrel test)', async () => {
        expect(
            YAML.stringify(
                await runCommand({
                    args: [
                        'chat',
                        '--context-repo',
                        'github.com/sourcegraph/sourcegraph',
                        '--show-context',
                        '-m',
                        'what is squirrel? Explain as briefly as possible.',
                    ],
                })
            )
        ).toMatchSnapshot()
    }, 20_000)

    it.skipIf(isWindows())(
        '--context-file (animal test)',
        async () => {
            expect(
                YAML.stringify(
                    await runCommand({
                        args: [
                            'chat',
                            '--context-file',
                            'animal.ts',
                            '--show-context',
                            '-m',
                            'implement a cow. Only print the code without any explanation.',
                        ],
                    })
                )
            ).toMatchSnapshot()
        },
        20_000
    )
})
