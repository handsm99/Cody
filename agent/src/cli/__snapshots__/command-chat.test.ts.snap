// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`--context-file (animal test) 1`] = `
"command: cody chat chat --context-file animal.ts --show-context -m implement a
  cow. Only print the code without any explanation.
exitCode: 0
stdout: >+
  > Context items:


  > 1. WORKING_DIRECTORY/animal.ts




  Here's the implementation of a cow based on the provided \`StrangeAnimal\` interface:


  \`\`\`typescript:animal.ts

  class Cow implements StrangeAnimal {
      makesSound(): 'coo' | 'moo' {
          return 'moo';
      }
  }

  \`\`\`

stderr: ""
"
`;

exports[`--message (hello world test) 1`] = `
"command: cody chat chat -m respond with "hello" and nothing else
exitCode: 0
stdout: |+
  hello

stderr: ""
"
`;
